### PR TITLE
scrfd 人脸检测器复现

### PR DESCRIPTION
**Pull request by RainFrost1**
scrfd2.5： Paddle版本wider face数据集（easy、medium、hard）：0.93607，0.9229，0.7717
torch版本 0.9359， 0.9200， 0.7773。由于Paddle.topk 实现跟Pytorch有diff，导致结果略有diff

### CODE REVIEW

**Review by heavengate — COMMENTED** [review]
`configs/scrfd`目录下加下README吧，加一下模型库~

File: ppdet/data/source/widerface_v2.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+from .dataset import DetDataset
+
+from ppdet.utils.logger import setup_logger
+
+logger = setup_logger(__name__)
+
+
+@register
+@serializable
+class WIDERFaceDataSetV2(DetDataset):
```
Comment (yghstill) [inline ppdet/data/source/widerface_v2.py]: 为什么不能复用WIDERFaceDataSet

File: ppdet/modeling/assigners/atss_assigner.py [hunk 2]
```diff
@@ -137,8 +138,10 @@ def forward(self,
         # 2. compute center distance between all anchors and gt, [B, n, L]
         gt_centers = bbox_center(gt_bboxes.reshape([-1, 4])).unsqueeze(1)
         anchor_centers = bbox_center(anchor_bboxes)
+        #  gt2anchor_distances = (gt_centers - anchor_centers.unsqueeze(0)) \
```
Comment (yghstill) [inline ppdet/modeling/assigners/atss_assigner.py]: 注释内容需删除

File: ppdet/optimizer.py [hunk 3]
```diff
@@ -152,6 +152,34 @@ def __call__(self, base_lr, step_per_epoch):
         return boundary, value
 
 
+@serializable
+class LinearWarmupV2(object):
```
Comment (yghstill) [inline ppdet/optimizer.py]: 同理，为什么不能兼容LinearWarmup，可否使用开关控制？

**Review by yghstill — COMMENTED** [review]
代码冲突待解决

File: ppdet/data/source/widerface_v2.py (hunk 1 above)
Comment (RainFrost1) [reply ppdet/data/source/widerface_v2.py]: 这里用的是retinaface版本的widerface，带了5个keypoints

File: ppdet/optimizer.py (hunk 3 above)
Comment (RainFrost1) [reply ppdet/optimizer.py]: lr的计算差异比较大，对齐原repo代码，所以就重新写了一个

File: ppdet/data/source/widerface_v2.py (hunk 1 above)
Comment (yghstill) [reply ppdet/data/source/widerface_v2.py]: 5个关键点的wider face可以参考之前静态图下的实现，是否加一个with_lmk就可以？https://github.com/PaddlePaddle/PaddleDetection/blob/release/2.4/static/ppdet/data/source/widerface.py#L45

Comment (RainFrost1) [reply ppdet/data/source/widerface_v2.py]: 测试了一下，是不行的。label.txt文件的格式不一致

File: ppdet/modeling/backbones/resnet.py [hunk 4]
```diff
@@ -443,7 +448,12 @@ def __init__(self,
                  return_idx=[0, 1, 2, 3],
                  dcn_v2_stages=[-1],
                  num_stages=4,
-                 std_senet=False):
+                 std_senet=False,
+                 stem_max_pool_ks=3,
```
Comment (yghstill) [inline ppdet/modeling/backbones/resnet.py:452]: 新增的参数，需在下方增加comment

File: ppdet/modeling/coders/distance_bbox_coder.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from ppdet.core.workspace import register
+from ppdet.modeling.bbox_utils import distance2bbox, bbox2distance
+
+__all__ = ['DistanceBBoxCoder']
+
+
+@register
+class DistanceBBoxCoder:
```
Comment (yghstill) [inline ppdet/modeling/coders/distance_bbox_coder.py]: 此类直接放在ppdet.modeling.bbox_utils即可，不必新建文件

Comment (yghstill) [inline ppdet/modeling/coders/distance_bbox_coder.py:8]: 此类直接放在ppdet.modeling.bbox_utils即可，不必新建文件

File: ppdet/modeling/heads/scrfd_head.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    for i in range(0, kps.shape[1], 2):
+        px = kps[:, i] - points[:, i % 2]
+        py = kps[:, i + 1] - points[:, i % 2 + 1]
+        if max_dis is not None:
+            px = paddle.clip(px, min=0, max=max_dis - eps)
+            py = paddle.clip(py, min=0, max=max_dis - eps)
+        preds.append(px)
+        preds.append(py)
+    return paddle.stack(preds, -1)
+    # kps[..., 0::2] -= points[..., 0].unsqueeze(axis=-1)
```
Comment (yghstill) [inline ppdet/modeling/heads/scrfd_head.py]: 删除注释

Comment (yghstill) [inline ppdet/modeling/heads/scrfd_head.py:61]: 删除注释

File: ppdet/modeling/heads/scrfd_head.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+            pos_bbox_targets = paddle.gather(
+                flatten_bbox_targets, pos_inds, axis=0)
+            pos_decode_bbox_pred = paddle.gather(
+                flatten_bboxes, pos_inds, axis=0)
+
+            pos_cls_pred = paddle.gather(
+                flatten_assigned_scores, pos_inds, axis=0)
+            weight_targets = pos_cls_pred.detach()
+            weight_targets = F.sigmoid(1 - weight_targets)
+            # weight_targets = F.sigmoid(weight_targets)
```
Comment (yghstill) [inline ppdet/modeling/heads/scrfd_head.py]: same as above

Comment (yghstill) [inline ppdet/modeling/heads/scrfd_head.py:427]: same as above

File: ppdet/data/source/widerface_v2.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (yghstill) [inline ppdet/data/source/widerface_v2.py]: widerface_v2.py与widerface.py需合并

Comment (yghstill) [inline ppdet/data/source/widerface_v2.py:1]: widerface_v2.py与widerface.py需合并

File: configs/scrfd/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+|:------------:|:--------:|:----:|:-------:|:-------:|:---------:|:----------:|:---------:|:--------:|
+| scrfd2.5g  | 640  |    4    | 640e     | 0.9361 / 0.9229 / 0.7717 | - | 2.6 |[link](https://paddledet.bj.bcebos.com/models/scrfd2_5g.pdparams) | [Configuration File](./scrfd_r50_pafpn_2_5.yml) |
+
+**Notes:**
+
+- All above models are trained on WIDERFace with 4 GPUs.
+
+- Evaluation on WIDERFace can be found [here](../face_detection/README.md).
+
+- Dataset: The retinaface version ofWIDERFace is used, which can be downloaded [here](https://drive.google.com/file/d/1UW3KoApOhusyqSHX96yEDRYiNkd3Iv3Z/view?usp=sharing).
```
Comment (yghstill) [inline configs/scrfd/README.md]: ... version of WIDERFace?

Comment (yghstill) [inline configs/scrfd/README.md:19]: ... version of WIDERFace?
